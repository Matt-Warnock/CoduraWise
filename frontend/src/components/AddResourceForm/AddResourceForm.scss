.add-resource {
  &__form {
    width: 30em;
    margin: 1em;
    display: flex;
    flex-direction: column;
    align-items: center;
  }

  &__input-container {
    margin: 5px 0 5px 0;
    display: flex;
    justify-content: space-around;
    align-items: center;
  }
  &__input-container-with-instruction {
    display: flex;
    flex-direction: column;
    .add-resource__instruction {
      margin-right: 2.4em;
      align-self: flex-end;
    }
  }
  &__label {
    width: 7em;
  }
  &__input {
    width: 22em;
    padding: 0.3em;
  }

  &__button {
    width: 10em;
    margin: 1em;
    padding: 0.7em;
    background-color: #1d3557;
    color: white;
    font-weight: bold;
    border: none;
    border-radius: 6px;
    align-self: center;
  }
}
