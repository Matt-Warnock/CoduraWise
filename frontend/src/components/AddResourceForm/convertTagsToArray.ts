const convertTagsToArray = (tagsString: string) => {
  return tagsString.split("#").splice(1);
};

export default convertTagsToArray;
