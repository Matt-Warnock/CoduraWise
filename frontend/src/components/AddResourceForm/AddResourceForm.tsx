import React, { useState } from "react";
import { useNavigate } from "react-router-dom";
import { MediaType } from "../../models/MediaTypes";
import { routerPaths } from "../../routes/paths";
import useResources from "../../store/hooks/useResources";
import MediaCheckBox from "../MediaCheckBox/MediaCheckBox";
import StarRatingInput from "../StarRatingInput/StarRatingInput";
import convertTagsToArray from "./convertTagsToArray";
import "./AddResourceForm.scss";

const AddResourceForm = () => {
  const initialState = {
    title: "",
    link: "",
    tags: "",
  };
  const [formValues, setFormValues] = useState(initialState);
  const { postNewResource } = useResources();
  const [averageRating, setRating] = useState("1");
  const [checkedMediaType, setCheckedMediaType] = useState<MediaType>("video");

  const navigate = useNavigate();

  const handleInputChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    const { name, value } = event.target;
    setFormValues({
      ...formValues,
      [name]: value,
    });
  };

  const handleSubmit = (event: React.FormEvent<HTMLFormElement>) => {
    event.preventDefault();

    const path = routerPaths.home;
    navigate(path);
    const body = {
      ...formValues,
      averageRating,
      mediaType: checkedMediaType,
      tags: convertTagsToArray(formValues.tags),
    };

    postNewResource(body);
  };

  return (
    <>
      <form className="add-resource__form" onSubmit={handleSubmit}>
        <div className="add-resource__input-container">
          <label className="add-resource__label" htmlFor="title">
            Title:
          </label>
          <input
            className="add-resource__input"
            type="text"
            name="title"
            pattern="^[A-Za-z]{1}(.\s?)*"
            autoFocus
            required
            title="Requires a title"
            value={formValues.title}
            onChange={handleInputChange}
          />
        </div>
        <div className="add-resource__input-container">
          <label className="add-resource__label" htmlFor="link">
            Resource Link:
          </label>
          <input
            className="add-resource__input"
            type="text"
            name="link"
            pattern="[(http(s)?):\/\/(www\.)?a-zA-Z0-9@:%._\+~#=]{2,256}\.[a-z]{2,6}\b([-a-zA-Z0-9@:%_\+.~#?&//=]*)"
            autoFocus
            required
            title="Please give a valid URL to the resource"
            value={formValues.link}
            onChange={handleInputChange}
          />
        </div>
        <div className="add-resource__input-container-with-instruction">
          <div className="add-resource__input-container">
            <label className="add-resource__label" htmlFor="tags">
              Tags:
            </label>
            <input
              className="add-resource__input"
              type="text"
              name="tags"
              pattern="(#\w{3,})+"
              autoFocus
              required
              title="Please append each tag with a '#' without spaces."
              value={formValues.tags}
              onChange={handleInputChange}
            />
          </div>
          <small className="add-resource__instruction">
            Enter tags with a prepending &quot;#&quot; without spaces.
          </small>
        </div>
        <MediaCheckBox setCheckedMediaType={setCheckedMediaType} />
        <StarRatingInput setRating={setRating} />
        <input
          className="add-resource__button"
          type="submit"
          value="Submit Resource"
        />
      </form>
    </>
  );
};

export default AddResourceForm;
