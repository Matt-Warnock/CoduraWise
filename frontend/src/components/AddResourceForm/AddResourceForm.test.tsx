import TestRenderer from "react-test-renderer";
import React from "react";
import ContextRouterMock from "../../mocks/contextRouterMock";
import AddResourceForm from "./AddResourceForm";

describe("given a resource form", () => {
  test("snapshot test: the html matches", () => {
    const resourceFormComponent = TestRenderer.create(
      <ContextRouterMock>
        <AddResourceForm />
      </ContextRouterMock>,
    );

    expect(resourceFormComponent.toJSON()).toMatchSnapshot();
  });
});

