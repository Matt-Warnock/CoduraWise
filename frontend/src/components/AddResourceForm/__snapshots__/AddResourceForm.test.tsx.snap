// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`given a resource form snapshot test: the html matches 1`] = `
<form
  className="add-resource__form"
  onSubmit={[Function]}
>
  <div
    className="add-resource__input-container"
  >
    <label
      className="add-resource__label"
      htmlFor="title"
    >
      Title:
    </label>
    <input
      autoFocus={true}
      className="add-resource__input"
      name="title"
      onChange={[Function]}
      pattern="^[A-Za-z]{1}(.\\\\s?)*"
      required={true}
      title="Requires a title"
      type="text"
      value=""
    />
  </div>
  <div
    className="add-resource__input-container"
  >
    <label
      className="add-resource__label"
      htmlFor="link"
    >
      Resource Link:
    </label>
    <input
      autoFocus={true}
      className="add-resource__input"
      name="link"
      onChange={[Function]}
      pattern="[(http(s)?):\\\\/\\\\/(www\\\\.)?a-zA-Z0-9@:%._\\\\+~#=]{2,256}\\\\.[a-z]{2,6}\\\\b([-a-zA-Z0-9@:%_\\\\+.~#?&//=]*)"
      required={true}
      title="Please give a valid URL to the resource"
      type="text"
      value=""
    />
  </div>
  <div
    className="add-resource__input-container-with-instruction"
  >
    <div
      className="add-resource__input-container"
    >
      <label
        className="add-resource__label"
        htmlFor="tags"
      >
        Tags:
      </label>
      <input
        autoFocus={true}
        className="add-resource__input"
        name="tags"
        onChange={[Function]}
        pattern="(#\\\\w{3,})+"
        required={true}
        title="Please append each tag with a '#' without spaces."
        type="text"
        value=""
      />
    </div>
    <small
      className="add-resource__instruction"
    >
      Enter tags with a prepending "#" without spaces.
    </small>
  </div>
  <fieldset
    className="media-selection"
  >
    <h3>
      Select Media Type:
    </h3>
    <div>
      <input
        defaultChecked={true}
        id="videomedia"
        name="mediaType"
        onChange={[Function]}
        type="radio"
        value="video"
      />
      <label
        htmlFor="videomedia"
      >
        video
      </label>
    </div>
    <div>
      <input
        defaultChecked={false}
        id="articlemedia"
        name="mediaType"
        onChange={[Function]}
        type="radio"
        value="article"
      />
      <label
        htmlFor="articlemedia"
      >
        article
      </label>
    </div>
    <div>
      <input
        defaultChecked={false}
        id="bookmedia"
        name="mediaType"
        onChange={[Function]}
        type="radio"
        value="book"
      />
      <label
        htmlFor="bookmedia"
      >
        book
      </label>
    </div>
    <div>
      <input
        defaultChecked={false}
        id="tutorialmedia"
        name="mediaType"
        onChange={[Function]}
        type="radio"
        value="tutorial"
      />
      <label
        htmlFor="tutorialmedia"
      >
        tutorial
      </label>
    </div>
    <div>
      <input
        defaultChecked={false}
        id="coursemedia"
        name="mediaType"
        onChange={[Function]}
        type="radio"
        value="course"
      />
      <label
        htmlFor="coursemedia"
      >
        course
      </label>
    </div>
  </fieldset>
  <div
    className="rate"
  >
    <input
      id="star5"
      name="rate"
      onChange={[Function]}
      type="radio"
      value="5"
    />
    <label
      htmlFor="star5"
      title="text"
    >
      5 stars
    </label>
    <input
      id="star4"
      name="rate"
      onChange={[Function]}
      type="radio"
      value="4"
    />
    <label
      htmlFor="star4"
      title="text"
    >
      4 stars
    </label>
    <input
      id="star3"
      name="rate"
      onChange={[Function]}
      type="radio"
      value="3"
    />
    <label
      htmlFor="star3"
      title="text"
    >
      3 stars
    </label>
    <input
      id="star2"
      name="rate"
      onChange={[Function]}
      type="radio"
      value="2"
    />
    <label
      htmlFor="star2"
      title="text"
    >
      2 stars
    </label>
    <input
      id="star1"
      name="rate"
      onChange={[Function]}
      type="radio"
      value="1"
    />
    <label
      htmlFor="star1"
      title="text"
    >
      1 star
    </label>
  </div>
  <input
    className="add-resource__button"
    type="submit"
    value="Submit Resource"
  />
</form>
`;
